module/VuFind/src/VuFind/Db/Service/ChangeTrackerService.php

```
@@ -0,0 +1,286 @@
+     * @param int    $change The timestamp of the last record change.
+     *
+     * @return ChangeTracker
+     */
+    public function index($core, $id, $change)
+    {
+        // Get a row matching the specified details:
+        $row = $this->retrieveOrCreate($core, $id);
+
+        // Flag to indicate whether we need to save the contents of $row:
+        $saveNeeded = false;
+        $utcTime = \DateTime::createFromFormat('U', $change, new \DateTimeZone('UTC'));
+        $now = new \DateTime('now', new \DateTimeZone('UTC'));
+        // Make sure there is a change date in the row (this will be empty
+        // if we just created a new row):
+        if (empty($row->getLastRecordChange())) {
+            $row->setLastRecordChange($utcTime);
+            $saveNeeded = true;
+        }
+
+        // Are we restoring a previously deleted record, or was the stored
+        // record change date before current record change date?  Either way,
+        // we need to update the table!
+        if (!empty($row->getDeleted()) || $row->getLastRecordChange() < $utcTime) {
+            // Save new values to the object:
+            $row->setLastIndexed($now);
+            $row->setLastRecordChange($utcTime);
+
+            // If first indexed is null, we're restoring a deleted record, so
+            // we need to treat it as new -- we'll use the current time.
+            if (empty($row->getFirstIndexed())) {
+                $row->setFirstIndexed($now);
+            }
+
+            // Make sure the record is "undeleted" if necessary:
+            $row->setDeleted(null);
+
+            $saveNeeded = true;
+        }
+
+        // Save the row if changes were made:
+        if ($saveNeeded) {
+            try {
+                $this->persistEntity($row);
+            } catch (\Exception $e) {
+                $this->logError('Could not update the change tracker table: ' . $e->getMessage());
+                return false;
```

Should we rethrow the exception in this situation, since this function is expected to return a ChangeTracker? Or do we need to adjust the return type definition to allow a false return?